src/internal/clusterstate/utils/pg_batcher.go

```diff
@@ -0,0 +1,186 @@
+
+Current limitations:
+   - Only UPDATE statements are supported.
+   - Only one column is supported in the WHERE clause.
+
+It is likely that we'll want to enhance this in the future to support INSERT and/or DELETE statements,
+and to allow for multiple columns in the WHERE clause. The functions have been written with this in mind
+to make them easy to expand upon.
+*/
+
+// row contains the values to be used for an insert/update of one row in Postgres.
+type row struct {
+	column_values []any // Contains the values to be inserted/updated for any given row
+	where_values  []any // Contains the values to be used in the WHERE clause
+}
+
+// postgresBatcher is an object to be used for caching rows of insert/update statements and sending them to Postgres in batches.
+type postgresBatcher struct {
+	tx         *pachsql.Tx // The transaction object
+	action     string      // One of the following: INSERT, UPDATE, DELETE (currently only UPDATE is implemented)
+	table      string      // The name of the table
+	columns    []string    // The list of columns whose values will be set by the query
+	w_columns  []string    // The list of columns used in the WHERE clause
+	set_string string      // The SET clause for the query (ex: "col1 = x.col1, col2 = x.col2")
+	rows       []row       // An array of row objects with values for insert/update/delete
+	max        int         // The maximum number of rows per batch
+	batch_num  int         // A count of the number of batches executed (for info logging only)
+}
+
+// NewPostgresBatcher creates a new batcher.
+func NewPostgresBatcher(tx *pachsql.Tx, action string, table string, columns []string, w_columns []string, batch_size int) (error, *postgresBatcher) {
+
+	// Validate the action; currently only UPDATE is supported
+	if !strings.EqualFold(action, "UPDATE") {
+		if !strings.EqualFold(action, "INSERT") && !strings.EqualFold(action, "DELETE") {
+			return errors.Errorf("%s is not a valid action (must be one of: INSERT, UPDATE, DELETE)"), nil
+		} else {
+			return errors.Errorf("INSERT and DELETE are not currently implemented"), nil
+		}
+	}
+
+	// Validate that at least 1 column and no more than 1 w_column have been provided.
+	if len(columns) == 0 {
+		return errors.Errorf("at least one column must be provided; none were given"), nil
+	}
+	if len(w_columns) > 1 {
+		return errors.Errorf("currently only 1 column may be used in the WHERE clause, but %d were given", len(w_columns)), nil
+	}
+
+	// Array where each element corresponds to the set string for one column (ex: "col1 = x.col1")
+	ss := make([]string, len(columns))
+
+	// Loop through the columns, creating the set string for each
+	for i, col := range columns {
+		if col == "proto" {
+			// It’s a code smell that newPostgresBatcher knows about
+			// decoding a proto column, while the caller of Add must
+			// know about encoding it.
+			ss[i] = fmt.Sprintf(`%s = decode(x.%s, 'hex')`, col, col)
```

Comment:
Why do these need to be hex encoded?  The driver can handle []byte values in placeholder bindings.

Reply:
Initially it was because we were building the UPDATEs as strings without placeholders, but there were other issues with that approach. Bob and I spent a lot of time trying different variations on the SQL statement before we got this one (with placeholders) working.  I did try to change it to send the []byte rather than a hex-encoded string, but it fails with the following message:

ERROR: invalid byte sequence for encoding "UTF8": 0x98 (SQLSTATE 22021)

Since the hex-encoding works, and we'd already spent a significant amount of time here, I decided to stick with that.

Reply:
Long story short, it doesn't seem to work.  It gives the following error: 

`ERROR: invalid byte sequence for encoding "UTF8": 0xec 0xa1 0xc1 (SQLSTATE 22021)`